### PR TITLE
perf(ts): reuse Reader whilst parsing records

### PR DESCRIPTION
**Pull request by AaronO**
Reduces `McapStreamReader` heap usage by ~25% and boosts throughput by ~30%.

This is both a refactor and perf boost, with more room for improvement, key changes:
- Removes StreamBuffer, hoisted into McapStreamReader
- Reuses Reader and DataView class across parse calls, only resetting them when necessary (e.g: append in McapStreamReader)
- Splits `parseRecord` into small scoped parsing functions, this itself is perf neutral (slightly positive) but facilitates future monomorphic fast paths
- Moves offsets tracking into Reader which is cleaner and faster

### Before

```
McapStreamReader
        3.48±0.03 op/s  Heap Used: 49.56±12.75 MB/op    Heap Total: 41.47±11.83 MB/op   ArrayBuffers: 112.95±6.87 MB/op
McapIndexedReader
        2.15±0.02 op/s  Heap Used: 70.02±2.84 MB/op     Heap Total: 58.34±3.36 MB/op    ArrayBuffers: 17.86±0.76 MB/op
McapIndexedReader_reverse
        2.18±0.01 op/s  Heap Used: 59.92±2.86 MB/op     Heap Total: 39.81±1.00 MB/op    ArrayBuffers: 14.58±1.42 MB/op
```

### After

```
McapStreamReader
        4.47±0.08 op/s  Heap Used: 42.35±2.23 MB/op     Heap Total: 32.93±3.76 MB/op    ArrayBuffers: 105.93±12.19 MB/op
McapIndexedReader
        2.38±0.02 op/s  Heap Used: 72.00±1.70 MB/op     Heap Total: 55.12±2.51 MB/op    ArrayBuffers: 17.86±1.85 MB/op
McapIndexedReader_reverse
        2.38±0.02 op/s  Heap Used: 63.41±1.55 MB/op     Heap Total: 39.33±0.53 MB/op    ArrayBuffers: 18.40±1.60 MB/op
```

### Followups

- Simplify parsing code further after exploring monomorphic callpaths
- Can further improve DataView/Reader churn in indexed readers

### CODE REVIEW

File: typescript/core/src/McapStreamReader.ts
Comment (jtbandes) [inline typescript/core/src/McapStreamReader.ts:97]: Curious what the motivation was for merging this logic into the McapStreamReader – it reduces encapsulation a bit, but maybe there is a perf win? (if so, do we know what the specific source of the win is?)

Comment (jtbandes) [inline typescript/core/src/McapStreamReader.ts:97]: nit: append() is so small I could see merging appendOrShift into it as a single function. Initially I was confused why `#reader` was not updated in appendOrShift, then I noticed it was here – that update could just be moved to the end of the function if they are combined

File: typescript/core/src/parse.ts
Comment (jtbandes) [inline typescript/core/src/parse.ts:31]: there's been a consensus (or ruling?) among the team that naked boolean parameters harm readability at call sites, so if this _needs_ to be this way for performance reasons, we'll need at least a comment to that effect. (I'm not sure how big of an effect the removal of the params object had?)

File: typescript/core/src/parse.ts
Comment (jtbandes) [inline typescript/core/src/parse.ts:47]: this behavior of resetting the reader is interesting and should probably be called out in the doc comment on this function, and can we add unit tests for it?

File: typescript/core/src/parse.ts
Comment (jtbandes) [inline typescript/core/src/parse.ts:105]: should this be more like an assert()? or is there some legitimate case today where this is not already at the right offset?

File: typescript/core/src/parse.ts
Comment (jtbandes) [inline typescript/core/src/parse.ts:190]: nit: could keep this as 2+4+8+8 for readability (I imagine/hope that would not make a difference in performance since it would be constant-folded?)

File: typescript/core/src/parse.ts
Comment (jtbandes) [inline typescript/core/src/parse.ts:208]: drive-by nit: I think this would be better named `recordsByteLength` since it's the length of the `records` field. As is, I was confused by how similar it is to the new parameter `recordLength`

File: typescript/core/src/parse.ts
Comment (jtbandes) [inline typescript/core/src/parse.ts:210]: nit: I find this condition harder to understand than the old version. Perhaps it can be flipped around more like `... > recordLength` as before?

File: typescript/core/src/parse.ts
Comment (jtbandes) [inline typescript/core/src/parse.ts:213]: I wonder if we should actually defer to the u8ArrayCopy method for bounds-checking? Now that the Reader class is being used for this, it feels redundant for the bounds check to be done by the caller

File: typescript/core/src/parse.ts
Comment (jtbandes) [inline typescript/core/src/parse.ts:286]: similar to above comment, it feels like we could now rely on u8ArrayCopy() for the bounds check

File: typescript/core/src/parse.ts
Comment (jtbandes) [inline typescript/core/src/parse.ts:296]: nit: might be a little simpler to capture `reader.offset` after reading the `expectedCrc` and then reset it to that value here

File: typescript/core/src/parse.ts
Comment (jtbandes) [inline typescript/core/src/parse.ts:274]: same comment about boolean parameters applies here

**Review by jtbandes — COMMENTED** [review]
Seems like a nice refactor overall! Most of it makes sense to me, I just left a few questions and cleanup suggestions. I do find the duplication of `reader.offset = startOffset + recordLength;` a bit repetitive and I wonder if there is a way that could be cleaned up...

File: typescript/core/src/Reader.ts
Comment (defunctzombie) [inline typescript/core/src/Reader.ts:117]: These methods could use a comment on what they do and how to use them.

File: typescript/core/src/Reader.ts
Comment (defunctzombie) [inline typescript/core/src/Reader.ts:19]: This could use a comment on what you are resetting and how it works, etc.

File: typescript/core/src/StreamBuffer.test.ts
Comment (defunctzombie) [reply typescript/core/src/StreamBuffer.test.ts:1]: Would it be covered by all the conformance testing? These tests seem really specific to this now deleted class so seems ok?

File: typescript/core/src/parse.ts
Comment (AaronO) [reply typescript/core/src/parse.ts:213]: Because throwable functions are harder for v8 to optimize / not-cost-free, I plan to shift error handling up not down the call tree.

In the optimistic/happy/common paths we should never throw here, so IMO it can be handled ad-hoc, e.g: return `null` then run ad-hoc code to determine precisely what the error was.

The indirection might be a little counterintuitive, but will help keep hot path call tree fast / simple

File: typescript/core/src/parse.ts
Comment (AaronO) [reply typescript/core/src/parse.ts:286]: As explained above, best to move error throwing up not down the tree

File: typescript/core/src/parse.ts
Comment (AaronO) [reply typescript/core/src/parse.ts:296]: Will do a cleanup pass on all parse logic once perf passes are mostly settled

File: typescript/core/src/parse.ts
Comment (achim-k) [inline typescript/core/src/parse.ts:111]: Shouldn't this be a `u8ArrayCopy` here? Or is the data kept internally?

File: typescript/core/src/parse.ts
Comment (achim-k) [inline typescript/core/src/parse.ts:142]: nit: Rename to `startOffset` to stay consistent with other `parse*` functions

File: typescript/core/src/parse.ts
Comment (achim-k) [inline typescript/core/src/parse.ts:171]: Would it performance-wise be better to avoid the closures? E.g. something like that:

```ts
function readString(reader: Reader) {
  return reader.string();
}
...

const metadata = reader.map(readString, readString)
```

Comment (AaronO) [reply typescript/core/src/parse.ts:171]: Yes, it'd likely be a small net gain, but it's worth optimizing only/mainly if metadata are in the hot path (like messages).

**Review by jtbandes — APPROVED** [review]
Looks good, but before merging please add a comment to explain the disabling of the no-boolean-parameters lint rule (see https://github.com/foxglove/mcap/pull/1212#discussion_r1715989185 and also consider this suggestion: https://github.com/foxglove/mcap/pull/1212#discussion_r1716715764)

**Pull request merged**
